### ISSUE TITLE
@emotion/styled: Add Transient Props

### ISSUE THREAD

**Issue by petermikitsh** [emotion-js/emotion#2193]
**The problem**

In 5.1.0, styled-components introduced [transient props](https://styled-components.com/releases#v5.1.0). Props prefixed with a dollar sign (`$`) are "mark[ed] as transient and styled-components knows not to add it to the rendered DOM element or pass it further down the component hierarchy".

**Proposed solution**

This would be useful functionality to support -- I am exploring migrating from `styled-components` to `@emotion/styled`.

**Alternative solutions**

None suggested. The intent is to follow this implementation detail introduced by another CSS-in-JS project.

**Additional context**

[Material UI](https://next.material-ui.com/guides/migration-v4/) v4 -> v5 is migrating to emotion, and my project uses styled-components today. I'm trying to get my codebase to coalesce around a single CSS in JS solution, and the fewer the API differences there are, the lower the barrier to migration is.

**Comment by Andarist** [comment]
I'm not super keen about it but we also maintain compatibility with SC (unless we feel strongly that it's not worth it) so it might be worth adding support for this. @mitchellhamilton thoughts?

**Comment by mitchellhamilton** [comment]
I'm also not super keen. For migration purposes, you could have a wrapper around styled that implements it with shouldForwardProp.

> we also maintain compatibility with SC (unless we feel strongly that it's not worth it) so it might be worth adding support for this

This isn't totally true. We don't have `.attrs`, `.withConfig`, `createGlobalStyle`, the `as` prop is forwarded by default for non-dom element types, `css` is eagerly evaluated and there are probably some other subtle differences. I think that anything you can do in s-c, there should be an equivalent way to do in Emotion, if it's not the exact same API though, that's fine.

**Comment by jmca** [comment]
One downfall of using a wrapper function:
When the `styled` tag is picked up by Emotion's babel `autoLabel`, the `[local]` name generated will be the name of the wrapper function (`StyledEmotionInterop` in the above case), and not the component defined.

**Comment by jmca** [comment]
This is the pattern I have been using successfully with Material UI:
```
export const transientOptions: Parameters<CreateStyled>[1] = {
  shouldForwardProp: (propName: string) => !propName.startsWith('$'),
};
```
```
const SomeComponent = styled(
  Typography,
  transientOptions,
)<{ $customProp: boolean }>(({ theme, $customProp }) => ({...}));
```

Note that the `shouldForwardProp` will apply to components that extend the root component:
```
const AnotherComponent = styled(
  SomeComponent  /* no need to pass transientOptions again */
)<{ $customProp2: boolean }>(({ theme, $customProp2 }) => ({...}));
```
The downfall is that it is opt-in (you have to explicitly use `transientOptions`), at least on the root component.

**Comment by Andarist** [comment]
> When the styled tag is picked up by Emotion's babel autoLabel, the [local] name generated will be the name of the wrapper function (StyledEmotionInterop in the above case), and not the component defined.

I'm not sure if that's totally correct - you should be able to utilize [`importMap`](https://github.com/emotion-js/emotion/blob/master/packages/babel-plugin/README.md#importmap) to make labels behave pretty much the same as with the "builtin" `styled`.

**Comment by jmca** [comment]
Oh nice. I actually didn't even know that existed. Guess I should look at those docs more 😅

**Comment by jmca** [comment]
@Andarist Could you give an example of `importMap`'s usage? The docs are a bit brief.

**Comment by Andarist** [comment]
https://github.com/emotion-js/emotion/blob/fa977675e1df05f210005ccd3461d6cdaf941b42/packages/babel-plugin/__tests__/import-mapping/import-mapping.js#L9-L38

**Comment by jmca** [comment]
@Andarist How does this apply to relative imports within the same codebase?

Say a file in the same codebase defines and exports a `styled` wrapper function.
Then another file (in the same codebase) imports that file via relative import path.
How would one configure `importMap` to make sure that wrapper function doesn't "steal" the auto-label?

**Comment by Andarist** [comment]
Well - I would recommend creating a package for your own styled wrapper so you could configure a single entry for the `importMap`. Relative paths makes things worse - although you could copy the same thing with different paths leading to your `styled` from your app..

> How would one configure importMap to make sure that wrapper function doesn't "steal" the auto-label?

I don't think it should steal it - but it might add an unnecessary label in this case. You could configure Babel to ignore the file with styled wrapper so it wouldn't happen. Wrapping in a package would also make it easier.

**Comment by jmca** [comment]
@Andarist Thanks for the suggestion. Too bad `importMap` doesn't support regex.

**Comment by mitchellhamilton** [comment]
> I'd like to hear more about the specifics of the concerns (e.g., is there a weak or strong objection to the implementation detail). Thanks!

It's mostly around the fact that it creates a distinction between "styled components" and "regular components" with the naming scheme when whether something is a styled component or not should just be an implementation detail. We also try to maintain that `styled(SomeStyledComponent)` is practically equivalent to `styled(forwardRef((props, ref) => <SomeStyledComponent {...props} ref={ref} />))`(which iirc we do with the exception of withComponent) but you would likely want to make that not true for transient props(and it isn't with SC: https://codesandbox.io/s/muddy-surf-ihu6e) which we'd like to avoid.

**Comment by sauldeleon** [comment]
> This is the pattern I have been using successfully with Material UI:
> 
> ```
> export const transientOptions: Parameters<CreateStyled>[1] = {
>   shouldForwardProp: (propName: string) => !propName.startsWith('$'),
> };
> ```
> 
> ```
> const SomeComponent = styled(
>   Typography,
>   transientOptions,
> )<{ $customProp: boolean }>(({ theme, $customProp }) => ({...}));
> ```
> 
> Note that the `shouldForwardProp` will apply to components that extend the root component:
> 
> ```
> const AnotherComponent = styled(
>   SomeComponent  /* no need to pass transientOptions again */
> )<{ $customProp2: boolean }>(({ theme, $customProp2 }) => ({...}));
> ```
> 
> The downfall is that it is opt-in (you have to explicitly use `transientOptions`), at least on the root component.

Hi,

just I want to say that I have used this solution and works great. Would be an option to add a helper function exportable from Emotion and thats all? It can be maybe customizable with a regex pattern maybe?

**Comment by callmeberzerker** [comment]
has there been any progress on this or any changes of opinion if it should be done? this is really needed for DX, plus styled-components API parity...

**Comment by Andarist** [comment]
The API parity with Styled Components is not a goal. The mentioned concerns expressed here are still very much true. It also doesn't seem like there are a lot of people looking for this functionality since this thread isn't that busy at the end (I would probably have expected it to be much more busy based on this being in SC and all). This also has an impact on our consumers and I would be wary about including support for this in the current major version.

**Comment by JanStevens** [comment]
A lot of people new to frontend code get bitten by this, rest spreading a bunch of props on a styled component, only to notice way to late that suddenly there are a bunch of additional properties on the HTML tag making it invalid, render ugly or other horrible things.

Transient props is an elegant solution, yea sure you can already do it now but its not convenient, passing around the same function for every styled component. Indeed since v5 release of material UI I expect a lot more people will get bitten by this eventually

**Comment by callmeberzerker** [comment]
> A lot of people new to frontend code get bitten by this, rest spreading a bunch of props on a styled component, only to notice way to late that suddenly there are a bunch of additional properties on the HTML tag making it invalid, render ugly or other horrible things.
> 
> Transient props is an elegant solution, yea sure you can already do it now but its not convenient, passing around the same function for every styled component. Indeed since v5 release of material UI I expect a lot more people will get bitten by this eventually

Exactly, and I don't understand the resistance to implement this functionality. This is the only real reason why I am still using `styled-components` (or `mui-styled-engine-sc`) vs the default engine in `material-ui@5.x` which is `emotion`. 

I don't wanna be bothered to create:

```
const TabsStyled = styled(MuiTabs, {
  shouldForwardProp: prop =>
    isPropValid(prop) && prop.startsWith('$')
})`
   color: ${props =>
    props.$myCustomVariantThatStartsWithDollarSign ? 'hotpink' : 'turquoise'};

`
```

And the last thing I need is yet another `styled` import (wrapper) in my codebase (`babel-macros` also don't work if the files are imported from elsewhere, but don't quote me on that).  I currently have this issue on my watchlist but I would want clear indication from `emotion` team if they would add this, or we should just settle on using `styled-components` if we want this kind of functionality out of the box.

**Comment by sauldeleon** [comment]
Hi @callmeberzerker I agree with you that this is a very shitty way, but it can be improved like

```tsx
export const MyComponent= styled('div', transientOptions)<StyledProps>`
  background-color: var(--surface-${props => props.$awesomeTheme});
```

having in some common file

```
import { CreateStyled } from '@emotion/styled'

export const transientOptions: Parameters<CreateStyled>[1] = {
  shouldForwardProp: (propName: string) => !propName.startsWith('$'),
}
```

Yes, it would be better with a default built in approach for emotion, but for the moment I use this way, and the code does not look too messy!

**Comment by maapteh** [comment]
Thank you @jmca in my case im using a third party UI library Chakra-UI and i now can prevent props from passing and throwing errors to me (propName needs to be lowerCase, value must be String etc). Codebase became much better now :)

<img width="617" alt="Screenshot 2021-11-03 at 09 33 25" src="https://user-images.githubusercontent.com/25961356/140029790-9da4e2fa-6f58-469e-8b75-14310f4fb930.png">

**Comment by WilliamHoangUK** [comment]
I've recently migrated a project I'm working on to MUI v5 and started using emotion for the first time. Came to this thread after noticing an error about receiving a boolean value for a non-boolean attribute and was hoping for a neat solution to deal with it. [Came across this for styled-components,](https://github.com/styled-components/styled-components/issues/1198#issuecomment-696306096) and was hoping for similar functionality for emotion. I echo the sentiments of others who don't want to have to explicitly filter out props that we don't want to pass to the DOM.

**Comment by fwextensions** [comment]
I came to this thread from the same SC issue as @WilliamHoangUK, having needed transient props for the first time after moving to MUI v5.  Having some sort of built-in solution would be preferable to the various workarounds. 

For now, I'm just passing `1` or `0` for a boolean `filled` prop, which suppresses the warning, but does cause a `filled="1"` attribute to show up on the DOM element.  Unsightly, but the easiest solution for this one case.

**Comment by jamesmfriedman** [comment]
Implementing a custom wrapper is a bit of a hack and its non obvious as a solution. This is an absurdly common problem. The current solutions as implemented using filter functions do not really account for the size of this issue. Unless I’m writing styles that are bespoke to me, I almost always have to pass a prop for some condition to styles, meaning I will have to implement this function the majority of the time.

**Comment by Archheretic** [comment]
Transient props would be appreciated.

**Comment by 0xalmanac** [comment]
I really would love transient props implemented

**Comment by julianoappelklein** [comment]
I wrapped the styled instance returned by emotion to extend it.
The extension basically adds a default shouldForwardProp.

**transient-styled.ts:**
```js
import _styled from '@emotion/styled';

const styled = ((component: any, config: any) => {
  config = { shouldForwardProp: (prop: string) => !prop.startsWith('$'), ...config };
  return _styled(component, config);
}) as (typeof _styled);

const tags = 'a|abbr|address|area|article|aside|audio|b|base|bdi|bdo|big|blockquote|body|br|button|canvas|caption|cite|code|col|colgroup|data|datalist|dd|del|details|dfn|dialog|div|dl|dt|em|embed|fieldset|figcaption|figure|footer|form|h1|h2|h3|h4|h5|h6|head|header|hgroup|hr|html|i|iframe|img|input|ins|kbd|keygen|label|legend|li|link|main|map|mark|marquee|menu|menuitem|meta|meter|nav|noscript|object|ol|optgroup|option|output|p|param|picture|pre|progress|q|rp|rt|ruby|s|samp|script|section|select|small|source|span|strong|style|sub|summary|sup|table|tbody|td|textarea|tfoot|th|thead|time|title|tr|track|u|ul|var|video|wbr|circle|clipPath|defs|ellipse|foreignObject|g|image|line|linearGradient|mask|path|pattern|polygon|polyline|radialGradient|rect|stop|svg|text|tspan'.split('|');
for(const tag of tags){
  (styled as any)[tag] = styled(tag as any);
}

export default styled;
```

Then you just have to import and use it:
```
import styled from './transient-styled';
```

And no more annoying props.
This could easily become a tiny npm package. Maybe emotion-style-transient

**Comment by jmca** [comment]
@julianoappelklein That is a valid approach, but a couple hiccups with Babel + some workarounds mentioned at:
- https://github.com/emotion-js/emotion/issues/2193#issuecomment-766168894
- https://github.com/emotion-js/emotion/issues/2193#issuecomment-766282405
- https://github.com/emotion-js/emotion/issues/2193#issuecomment-766322021

**Comment by julianoappelklein** [comment]
Thanks @jmca - I see you mentioned earlier that this solution generates a different label.

> One downfall of using a wrapper function:
> When the styled tag is picked up by Emotion's babel autoLabel, the [local] name generated will be the name of the wrapper function (StyledEmotionInterop in the above case), and not the component defined.

I confess I don't understand the implications of this. I'm not seeing any side effect. Can you elaborate more?

**Comment by pkaufi** [comment]
I agree that having this implemented would make my life much easier. We're also just in process of migrating from SC to emotion, and this is a huge pain point. If you don't want to support it out of the box, maybe a configuration could be added, where it's possible to add a default behaviour for `shouldForwardProp`?

**Comment by jongbelegen** [comment]
> I agree that having this implemented would make my life much easier. We're also just in process of migrating from SC to emotion, and this is a huge pain point. If you don't want to support it out of the box, maybe a configuration could be added, where it's possible to add a default behaviour for `shouldForwardProp`?

I also would prefer overwriting the default behavior for shouldForwardProp, maybe through provider context or instance creation. We could document a clean way of having "styled-components" like transient props in emotion with this approach.

**Comment by mbcod3** [comment]
Why does emotion forward boolean prop to dom element by default? If it doesnt big use case for transient props would be resolved.

**Comment by sshmaxime** [comment]
I use Material UI (MUI) and it took me quite some time to realize that it was coming from emotion (being the default UI engine of MUI). I had to change the UI engine of MUI back to Styled-Component. This is how to do it: https://mui.com/material-ui/guides/styled-engine/#how-to-switch-to-styled-components

**Comment by KevinKra** [comment]
TLDR emotion will probably never add this feature 👎🏻

**Comment by Monstarrrr** [comment]
As a more junior developer this is holding me back quite a lot for something as basic and important as passing props around.

**Comment by Slapbox** [comment]
> The mentioned concerns expressed here are still very much true. It also doesn't seem like there are a lot of people looking for this functionality since this thread isn't that busy at the end

@Andarist I feel like this should be revisited, even if there's not a clear path forward. It's by far the most thumbs-upped issue on the tracker at this point with 51 currently - more than double the next highest.

**Comment by Ethorsen** [comment]
We all hope that some sort of mechanism will be implemented to help alleviate this problem.   But after a few years,  it just feels like the dev team wont do anything about it

**Comment by AlexSapoznikov** [comment]
Was getting another check on this thread after Oct 7, 2021. It's sad that this is still a problem, so avoiding emotion-js this time.

**Comment by Slapbox** [comment]
8 more thumbs-ups since my last comment about 75 days ago.

**Comment by ndeviant** [comment]
Any updates on this?

**Comment by mlnarik3pg** [comment]
We would like to see this feature!

**Comment by MadPeachTM** [comment]
so, it has been 2 years since the start of this conversation... has emotion implemented this feature? Docs on their website don't work and i can't find an answer anywhere

**Comment by Ethorsen** [comment]
@MadPeachTM Most probably not

answer is right here, simple JS version: 

```
export const withTransientProps = {
  shouldForwardProp: (propName) => propName !== 'theme' && !propName.startsWith("$"),
};
```

Then add it when necessary

```
const MyStyledStuff = styled(Stuff, withTransiantProps)(({ $someTransiantProp }) => ({
}));
```

**Comment by ValentynaTemy** [comment]
Any plans to add this?

**Comment by ed-peretokin** [comment]
Every solution on making a wrapper around the `styled` will be limited since TS will never know, how to process the overload or inferred generic types. I have made a solution, that will return us the original `styled` with using native Proxy functionality:
```typescript
import { CreateMUIStyled, styled as _styled, Theme } from '@mui/material/styles';

const createProxyStyled = (): CreateMUIStyled<Theme> => {
  const handler: ProxyHandler<CreateMUIStyled<Theme>> = {
    apply: function (target, _, argArray) {
      return target(
        argArray[0],
        {
          ...(argArray[1] || {}),
          shouldForwardProp: (prop) =>
            !prop.toString().startsWith('$') &&
            (argArray[1]?.shouldForwardProp(prop) || true),
        }
      );
    },
  };

  return new Proxy(_styled, handler);
};

export const styled = createProxyStyled();
```

**Comment by elektronik2k5** [comment]
Another possible workaround is monkey patching `@emotion/is-prop-valid` in your repo using `[patch-package](https://www.npmjs.com/package/patch-package)` (or the equivalent built-in feature in some package managers, like `yarn`). The patch would prepend `prop.startsWith('$') && ` to what `@emotion/is-prop-valid` does. It is ugly, gnarly, fragile across version updates and requires non trivial knowledge, but it should work.

I haven't tried this approach cause for me it is not a practical issue right now.

I've been using and recommending Emotion for a long time and it is sad to see that when the greater ecosystem (MUI) is choosing to embrace Emotion, the users' first experience is hitting this brick wall, which doesn't have an easy, simple and reliable workaround.

I understand and empathize with the maintainers' stance on API surface, scope and all that. Really, I do and I definitely support the notion that Emotion is not meant to be a drop-in replacement for styled-components, not foster its design choices.
However, it is also hard to ignore the fact that technically it should be a trivial issue to fix. @Andarist and @emmatown, if you don't wanna change the default, I suggest we come up with an easier escape hatch to make the migration smoother and allow easier customization - which also aligns well with Emotion's design goals (IIUC).

How about allowing overriding the default `shouldForwardProp` via a global configuration flag? For example, could we do it as part of the customizations which `CacheProvider` supports? Seems to me like the natural place for such an escape hatch.

**Comment by Touseef-ahmad** [comment]
I recently faced an issue, where I was getting warnings on the console, after hours of debugging I found the problem was that the props were being passed from my styled component to the Mui Typography component, which it not ideal, we would like to have a way to pass the props through so we don't get these annoying warnings.

<img width="439" alt="Screenshot 2024-05-30 at 11 03 15 AM" src="https://github.com/emotion-js/emotion/assets/43235697/d5ca3da1-c582-4e4a-8b0e-7104061aebf9">

**Comment by oigewan** [comment]
> > I'd like to hear more about the specifics of the concerns (e.g., is there a weak or strong objection to the implementation detail). Thanks!
> 
> It's mostly around the fact that it creates a distinction between "styled components" and "regular components" with the naming scheme when whether something is a styled component or not should just be an implementation detail. We also try to maintain that `styled(SomeStyledComponent)` is practically equivalent to `styled(forwardRef((props, ref) => <SomeStyledComponent {...props} ref={ref} />))`(which iirc we do with the exception of withComponent) but you would likely want to make that not true for transient props(and it isn't with SC: https://codesandbox.io/s/muddy-surf-ihu6e) which we'd like to avoid.

@emmatown 

I don't understand why allowing customization of the `isPropValid` function through context or something would break the pattern you describe.  At least one prop is already filtered (theme), so filtering props which are internal implementation details of the lib is already an accepted pattern.  We'd like a way to accomplish the same thing in a way that has a good DX.  Seems quite unfortunate that this feature request is being ignored as there appears to be quite a bit of desire for it.

**Comment by marano** [comment]
The lack of this feature is forcing us to migrate our app to using styled components instead. I don't really understand why this is so controversial. It is really weird setting up a new react app with emotion and then getting a bunch of react warnings about these invalid props being passed to the DOM.

**Comment by Nantris** [comment]
I don't get the sense Emotion is maintained anymore. I think we're all headed for `styled-components` in due time.

**Comment by jove4015** [comment]
Are we really all switching to styled-components though?  The MUI docs seem to indicate [exactly the opposite](https://mui.com/material-ui/integrations/styled-components/) (see the giant red warning about SSR at the top).

I am in the process of migrating from styled-components and hit this.  We relied on the transient props extensively and now it's a migration problem for us.  Nothing's actually broken in the app but we're getting tons of warnings and it's frustrating to have to fix these all individually.  I would just add to the list of people who think this is a convenience / DX feature that has improved their workflow and wish they still had.  I'm fine with emotion and honestly I think our app is working better with it, I just want this feature too.

**Comment by antonina-nesmelova** [comment]
Same issue. I'm migrating from styled-components because of bad performance.

**Comment by maapteh** [comment]
Unbelievable, something so simple now has so much effects. For example I add a prop to MUI and get it in the dom While a simple standard $ would have prevented this. I just came here and saw my own helper function https://github.com/emotion-js/emotion/issues/2193#issuecomment-958743435 which in MUI i can not do so easily since im creating the components during the theme creation.

**Comment by ilia-os** [comment]
As styled-components seems to be [fading](https://opencollective.com/styled-components/updates/thank-you) away, I believe now its even more important to add an opt-in into transient props starting with `$` to simplify styled-components -> emotion/styled transition

**Comment by fxdave** [comment]
@ilia-os Emotion is also fading away. In fact, every non-static CSS-IN-JS. Next.js users cannot use it, because it doesn't work with server components. It seems nextjs is quite popular to dictate the trends.

**Comment by Nantris** [comment]
I would hardly call slower growth "fading away" in either case. Server components are extremely far from being the only use case. Mobile and desktop apps have little reason not to prefer CSS in JS besides the small performance runtime impact.

**Comment by Andarist** [comment]
We don't plan to implement this feature as we don't believe in it. I understand that people are using this and relying it in their apps but it is certainly a *choice* they have made. SC and Emotion are 2 different libraries - even if they share the core of the API. There are more differences between them than just this though. They are not meant to be interchangeable - they never have been.

There are ways to work around this - like the mentioned `shouldForwardProp` API. This is exactly the kind of customization you might want to introduce in your codebase using this API.

**Comment by jove4015** [comment]
That's disappointing - due to [this bug](https://github.com/vercel/next.js/issues/46973#issuecomment-2682479509), the "shouldForwardProp" API is hard to use in a consistent way - if you want to share the functions passed to "shouldForwardProp" between modules, you constantly have to spread those options out into new objects - and if you forget to do this, your code will break (but only in certain environments, leading you not to notice it's broken).  Typescript, eslint - nothing notices anything is wrong until you're actually running the app and hit a wall of error messages.

With the $ approach, we never had any issue with managing which props get forwarded to a component.  With "shouldForwardProp", we constantly have to make sure we're doing it the right way to avoid surprise compilation errors.  I even had to write a github action script just to check for bad implementations of this and flag them.

Personally, I don't believe in making developers jump through all these hoops.

**Comment by Andarist** [comment]
@jove4015 I'd recommend wrapping `styled` and applying a custom `shouldForwardProp` there instead of configuring it at each call site.

**Comment by jove4015** [comment]
We have had a few developers make substantial use of the "$" functionality when we were using styled-components, and so we are passing shouldForwardProp in ~110 separate components.  Anywhere in which the same file is also making use of component selectors, this problem comes up (even if the component selectors aren't being used together with shouldForwardProp directly).  It is not a daily hassle once you get used to it (especially with a CI check in place), and had we known this was going to be an issue when we were moving things over to emotion, we probably would have done what @Andarist recommended and just set up a custom "styled".  I guess I was just trying to illustrate why I was still following this issue and waiting on a fix.

**Comment by elektronik2k5** [comment]
> [@jove4015](https://github.com/jove4015) I'd recommend wrapping `styled` and applying a custom `shouldForwardProp` there instead of configuring it at each call site.

In my experience, wrapping Emotion kills source maps cause then the source is always the wrapper. And source maps are one of the core features of why I choose to use Emotion in the first place.

**Comment by Andarist** [comment]
Source maps should work exactly the same with wrapped Emotion if you are using https://emotion.sh/docs/@emotion/babel-plugin#importmap . If they don't, please provide a repro case for further investigation.

**Comment by zbycz** [comment]
Hi - here is a custom typescript-eslint rule, which checks all `styled` call. If it has props, it asks for second parameter:

https://gist.github.com/zbycz/dee72fd0c4671c0e8e3e30de295a6df8


This helped me get rid of all the warnings in next.js like:
```
hydration-error-info.ts: Warning: React does not recognize the `$isMobileMode` prop on a DOM element. 
If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `$ismobilemode` instead. 
If you accidentally passed it from a parent component, remove it from the DOM element. Error Component Stack
    at div (<anonymous>)
    at eval (emotion-element-7a1343fa.browser.development.esm.js:48:27)
```

**Comment by woodreamz** [comment]
I am really surprised the transient props or a way to override the `shouldForwardProp` globally was never implemented. I don't think I am the only one but this bug is a huge NO GO to migrate to emotion :/
